Framework/Algorithms/src/MaskNonOverlappingBins.cpp

```diff
@@ -0,0 +1,136 @@
+
+/// Algorithm's category for identification. @see Algorithm::category
+std::string const MaskNonOverlappingBins::category() const {
+  return "Transforms\\Masking";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+std::string const MaskNonOverlappingBins::summary() const {
+  return "Marks bins in InputWorkspace which are out of the X range of the "
+         "second workspace.";
+}
+
+/// Return a list of the names of related algorithms.
+std::vector<std::string> const MaskNonOverlappingBins::seeAlso() const {
+  return {"MaskBins", "MaskBinsIf"};
+}
+
+/** Initialize the algorithm's properties.
+ */
+void MaskNonOverlappingBins::init() {
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::INPUT_WS, "", Kernel::Direction::Input),
+                  "A workspace to mask.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+                  "The masked workspace.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::COMPARISON_WS, "", Kernel::Direction::Input),
+                  "A workspace to compare the InputWorkspace's binning to.");
+  declareProperty(Prop::MASK_PARTIAL, false,
+                  "If true, mask also bins that overlap only partially.");
+}
+
+/// Returns a map from property name to message in case of invalid values.
+std::map<std::string, std::string> MaskNonOverlappingBins::validateInputs() {
+  std::map<std::string, std::string> issues;
+  API::MatrixWorkspace_const_sptr inputWS = getProperty(Prop::INPUT_WS);
+  API::MatrixWorkspace_const_sptr comparisonWS =
+      getProperty(Prop::COMPARISON_WS);
+  if (inputWS->getNumberHistograms() != comparisonWS->getNumberHistograms()) {
+    issues[Prop::COMPARISON_WS] =
+        "The number of histogams mismatches with " + Prop::INPUT_WS;
+  }
+  if (!inputWS->isHistogramData()) {
+    issues[Prop::INPUT_WS] =
+        "The workspace contains point data, not histograms.";
+  }
+  if (!comparisonWS->isHistogramData()) {
+    issues[Prop::COMPARISON_WS] =
+        "The workspace contains point data, not histograms.";
+  }
+  return issues;
+}
+
+/** Execute the algorithm.
+ */
+void MaskNonOverlappingBins::exec() {
+  API::MatrixWorkspace_sptr inputWS = getProperty(Prop::INPUT_WS);
+  API::MatrixWorkspace_sptr outputWS = getProperty(Prop::OUTPUT_WS);
+  if (inputWS != outputWS) {
+    outputWS = inputWS->clone();
+  }
+  API::MatrixWorkspace_const_sptr comparisonWS =
+      getProperty(Prop::COMPARISON_WS);
+  bool const maskPartial = getProperty(Prop::MASK_PARTIAL);
+  auto const nHist = inputWS->getNumberHistograms();
+  API::Progress progress(this, 0., 1., nHist);
+  // Tried to parallelize this loop but the performance test showed
+  // regression.
+  for (size_t histogramIndex = 0; histogramIndex < nHist; ++histogramIndex) {
```

Comment:
You could probably loop over the spectrum in parallel. But I guess there's no good performance reason to do this?

Reply:
Did you consider trying to determine if all X values were the same (across every spectra) before looping over 1 by 1 over spectra? I agree that the following would work and would be required in the general case I just wondered if there would be a simple way of establishing if the x-values were identical and therefore being able to apply the same masking.

Reply:
As the comment above the `for` loop states, parallelization led to poorer performance. If I remember correctly, the performance test for histogram data ran ~2x longer with my 8 core machine with `PARALLEL_FOR_IF`.

I did not consider the case of common X over all spectra. It sounds like a nice idea. I'll investigate.